ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -162,18 +307,52 @@ public boolean equals(Object o) {
     if (!version.equals(blobId.version)) {
       return false;
     }
-    if (!partitionId.equals(blobId.partitionId)) {
+    if (!flag.equals(blobId.flag)) {
       return false;
     }
-    if (!uuid.equals(blobId.uuid)) {
+    if (!datacenterId.equals(blobId.datacenterId)) {
       return false;
     }
-
-    return true;
+    if (!accountId.equals(blobId.accountId)) {
+      return false;
+    }
+    if (!containerId.equals(blobId.containerId)) {
+      return false;
+    }
+    if (!partitionId.equals(blobId.partitionId)) {
+      return false;
+    }
+    return uuid.equals(blobId.uuid);
   }
 
   @Override
   public int hashCode() {
     return Utils.hashcode(new Object[]{version, partitionId, uuid});
   }
+
+  /**
+   * Populate the missing fields with legacy values, and determine the version of format to serialize. Only
+   * when all the four fields of {@code flag}, {@code datacenterId}, {@code accountId}, and {@code containerId}
+   * are available, the serialization format version will be set to {@link #BLOB_ID_V2}, otherwise it will be
+   * {@link #BLOB_ID_V1}.
+   */
+  private void populateMissingFieldsAndDetermineVersion() {
+    version = flag == null && accountId == null && containerId == null && datacenterId == null ? BlobId.BLOB_ID_V1
```

Comment:
But the tricky part here is the index roll over. during this deployment, during replication some servers might send blobIds in V1 and some in blobIds in V2. If one server receives interleaved blobId  versions from same or multiple server, this will cause roll over of indexes for every size diff between successive storeKey(aka blobIds).